### PR TITLE
test: add integration tests for Matcher and Deal Rewards

### PR DESCRIPTION
**Pull request by justchillinghere**
1. Several tests for matcher covering logic of reserving all CUs from Peers provided
2. Test for checking effector filtering logic
3. Additional basic integration tests
4. Helpers and small fixes

### CODE REVIEW

File: test/core/market/Matcher.t.sol [hunk 1]
```diff
@@ -50,46 +53,223 @@ contract MatcherTest is TestWithDeployment {
...
+                if (deployment.diamondAsMarket.getComputeUnit(unitIds[i][j][k]).deal != address(0)) {
+                    continue;
+                }
+
+                // Create a Deal to be matched with the CU
+                unitIdsToMatcher[0][0] = unitIds[i][j][k];
+                DealMock deal = new DealMock(args);
+
+                // Check emiting events
+                vm.expectEmit(true, true, true, true, address(deployment.diamond));
```
Comment (crystalbit) [inline test/core/market/Matcher.t.sol:243]: true, false, false, true as only first event parameter is indexed

File: test/core/market/Matcher.t.sol [hunk 2]
```diff
@@ -162,55 +348,631 @@ contract MatcherTest is TestWithDeployment {
...
                     if (chosenComputeUnits == args.targetWorkers) {
                         break;
                     }
-                    unitIds2D[i][currentUnitIdx] = unitIds[i][j][k];
+                    unitIds2D[i][currentUnitIdx] = registeredOffers.unitIds[i][j][k];
                     currentUnitIdx += 1;
                     chosenComputeUnits += 1;
 
+                    // Current Protocol restricts more than 1 CU per Peer. So, the current test will fail for cases when params.unitCountPerPeer > 1.
                     vm.expectEmit(true, true, true, true, address(deployment.diamond));
```
Comment (crystalbit) [inline test/core/market/Matcher.t.sol:364]: could you fix here pls too: `vm.expectEmit(true, false, false, true,`

File: test/core/market/Matcher.t.sol [hunk 3]
```diff
@@ -162,55 +348,631 @@ contract MatcherTest is TestWithDeployment {
...
+            uint256 currentUnitIdx = 0;
+            for (uint256 j = 0; j < registeredOffers.peerIds[i].length; j++) {
+                for (uint256 k = 0; k < registeredOffers.unitIds[i][j].length; k++) {
+                    unitIds2D[i][currentUnitIdx] = registeredOffers.unitIds[i][j][k];
+                    currentUnitIdx += 1;
+
+                    // Current Protocol restricts more than 1 CU per Peer.
+                    // Check emiting events only for the first CU from the Peer.
+                    if (k == 0) {
+                        vm.expectEmit(true, true, true, true, address(deployment.diamond));
```
Comment (crystalbit) [inline test/core/market/Matcher.t.sol:581]: `vm.expectEmit(true, false, false, true,`

File: test/core/market/Matcher.t.sol [hunk 4]
```diff
@@ -162,55 +348,631 @@ contract MatcherTest is TestWithDeployment {
...
+                            args.creationBlock,
+                            args.appCID
+                        );
+                    }
+                }
+            }
+        }
+        deployment.diamondAsMarket.matchDeal(IDeal(address(deal)), registeredOffers.offerIds, unitIds2D);
+
+        // targetWorkers is set to be always leq than the total CUs available. So, deal is expected to be fully matched .
```
Comment (crystalbit) [inline test/core/market/Matcher.t.sol:595]: leq - typo?

File: test/core/market/Matcher.t.sol [hunk 5]
```diff
@@ -162,55 +348,631 @@ contract MatcherTest is TestWithDeployment {
...
+                new bytes32[](customParamsMatcher.peerCountPerOffer * customParamsMatcher.unitCountPerPeer);
+            for (uint256 j = 0; j < customParamsMatcher.peerCountPerOffer * customParamsMatcher.unitCountPerPeer; j++) {
+                uint256 peerIndex = j / customParamsMatcher.unitCountPerPeer;
+                uint256 unitIndex = j % customParamsMatcher.unitCountPerPeer;
+
+                unitIdsToMatcher[i][j] = registeredOffers[registerIndex].unitIds[offerIndex][peerIndex][unitIndex];
+
+                // Expect ComputeUnitMatched event for the first CU of each peer in offers with special effector
+                if (unitIndex == 0 && registerIndex == 1) {
+                    vm.expectEmit(true, true, true, true, address(deployment.diamond));
```
Comment (crystalbit) [inline test/core/market/Matcher.t.sol:927]: same pls

File: test/deal/Rewards.t.sol [hunk 6]
```diff
@@ -26,115 +26,474 @@ import {IDeal} from "src/deal/interfaces/IDeal.sol";
...
     using TestHelper for TestWithDeployment.Deployment;
 
-    struct Params {
-        uint256 minWorkers;
-        uint256 maxWorkersPerProvider;
-        uint256 targetWorkers;
-        uint256 pricePerWorkerEpoch;
-        uint256 minAmount;
-        uint256 protocolVersion;
+    event RewardWithdrawn(bytes32 computeUnitId, uint256 reward);
```
Comment (crystalbit) [inline test/deal/Rewards.t.sol:35]: you can usebetter use ICapacity.RewardWithdrawn, not to redeclare

**Review by crystalbit — APPROVED** [review]

**Comment by justchillinghere** [comment]
1. Added tests for Capacity Commitment rewards and slashing logic
2. Refactored TestHelper functions for better maintainability

File: test/core/capacity/CapacityCommitment.t.sol [hunk 7]
```diff
@@ -422,10 +430,19 @@ contract CapacityCommitmentTest is TestWithDeployment {
...
         bytes32[] memory unitIds = deployment.diamondAsMarket.getComputeUnitIds(peerId);
 
+        vm.expectEmit(false, false, false, true, address(deployment.diamondAsCapacity));
+        emit ICapacity.CommitmentStatsUpdated(
+            commitmentId, unitCount * ccDuration, unitCount, 0, 0, deployment.diamondAsCore.currentEpoch() - 1
+        );
+
         deployment.diamondAsCapacity.removeCUFromCC(commitmentId, unitIds);
+
+        vm.expectEmit(true, false, false, true, address(deployment.diamondAsCapacity));
```
Comment (crystalbit) [inline test/core/capacity/CapacityCommitment.t.sol:443]:
```suggestion
        vm.expectEmit(true, false, false, false, address(deployment.diamondAsCapacity));
```

File: test/core/capacity/CapacityCommitment.t.sol [hunk 8]
```diff
@@ -18,6 +18,7 @@
  */
 pragma solidity ^0.8.19;
 
+import "forge-std/console.sol";
```
Comment (crystalbit) [inline test/core/capacity/CapacityCommitment.t.sol:21]: delete pls

File: test/core/capacity/CapacityCommitment.t.sol [hunk 9]
```diff
@@ -435,39 +452,560 @@ contract CapacityCommitmentTest is TestWithDeployment {
...
+     * - Correct state transitions of the commitment
+     * - Accurate reward calculation based on submitted proofs
+     * - Proper reward distribution according to the delegation rate
+     * - Correct handling of vesting periods for reward withdrawal
+     */
+    function test_WitdrawRewardsAfterEndWithProofs() public {
+        uint256 rewardDelegationRate = 30;
+        uint256 ccDurationForTest = ccDuration * 20;
+
+        // StdCheats.skip(deployment.diamondAsCore.epochDuration() * 100);
```
Comment (crystalbit) [inline test/core/capacity/CapacityCommitment.t.sol:512]: unneeded comment?

File: test/core/market/Matcher.t.sol [hunk 10]
```diff
@@ -18,6 +18,7 @@
  */
 pragma solidity ^0.8.19;
 
+import "forge-std/console.sol";
```
Comment (crystalbit) [inline test/core/market/Matcher.t.sol:21]: delete pls

File: test/core/market/Matcher.t.sol [hunk 11]
```diff
@@ -29,275 +30,647 @@ import {IMarket} from "src/core/interfaces/IMarket.sol";
...
-                    break;
-                }
-                for (uint256 k = 0; k < unitIds[i][j].length; k++) {
-                    if (chosenComputeUnits == args.targetWorkers) {
-                        break;
-                    }
-                    unitIds2D[i][currentUnitIdx] = unitIds[i][j][k];
-                    currentUnitIdx += 1;
-                    chosenComputeUnits += 1;
+        deployment.diamondAsMarket.matchDeal(IDeal(address(deal)), registeredOffers.offerIds, unitIds2D);
```
Comment (crystalbit) [inline test/core/market/Matcher.t.sol:212]:
```suggestion
        deployment.diamondAsMarket.matchDeal(deal, registeredOffers.offerIds, unitIds2D);
```

File: test/core/market/Matcher.t.sol [hunk 12]
```diff
@@ -29,275 +30,647 @@ import {IMarket} from "src/core/interfaces/IMarket.sol";
...
-        for (uint256 i = 0; i < args.effectors.length; i++) {
-            _effectors.push(args.effectors[i]);
+                    bytes32[] memory offerIds = new bytes32[](1);
+                    offerIds[0] = registeredOffers.offerIds[i];
+
+                    bytes32[][] memory unitIds = new bytes32[][](1);
+                    unitIds[0] = new bytes32[](1);
+                    unitIds[0][0] = registeredOffers.unitIds[i][j][k];
+
+                    deployment.diamondAsMarket.matchDeal(IDeal(address(deal)), offerIds, unitIds);
```
Comment (crystalbit) [inline test/core/market/Matcher.t.sol:404]:
```suggestion
                    deployment.diamondAsMarket.matchDeal(deal, offerIds, unitIds);
```

File: test/deal/Rewards.t.sol [hunk 13]
```diff
@@ -24,117 +24,454 @@ import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
...
+            deal.setWorker(unitIdToCheck, workerId);
+        }
     }
 
     function setUp() public {
         _deploySystem();
+        // Transfer funds to provider and dealMaker
+        uint256 transferAmount = 10000 ether;
+        deployment.tUSD.transfer(dealMaker, transferAmount);
+        deal(address(dealMaker), transferAmount);
```
Comment (crystalbit) [inline test/deal/Rewards.t.sol:52]:
```suggestion
        vm.deal(address(dealMaker), transferAmount);
```

File: test/deal/Rewards.t.sol [hunk 14]
```diff
@@ -24,117 +24,454 @@ import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
...
         );
 
-        address computeProvider = address(0x1234567890123456789012345678901234567890);
-        bytes32 unitId = TestHelper.pseudoRandom(abi.encode(computeProvider, "unitId"));
-        bytes32 peerId = TestHelper.pseudoRandom(abi.encode(computeProvider, "peerId"));
-        bytes32 workerId = TestHelper.pseudoRandom(abi.encode(computeProvider, "workerId"));
-        bytes32 offerId = TestHelper.pseudoRandom(abi.encode(computeProvider, "offerId"));
+        // Match the Deal with the Offers. Only 1 CU is needed to be matched
+        bytes32[][] memory unitIdsByOffers = TestHelper.flattenUnitIdsToOffers(registeredOffers);
+        deployment.diamondAsMarket.matchDeal(IDeal(address(deal)), registeredOffers.offerIds, unitIdsByOffers);
```
Comment (crystalbit) [inline test/deal/Rewards.t.sol:106]:
```suggestion
        deployment.diamondAsMarket.matchDeal(deal, registeredOffers.offerIds, unitIdsByOffers);
```

File: test/deal/Rewards.t.sol [hunk 15]
```diff
@@ -24,117 +24,454 @@ import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
...
+        deployment.tUSD.forceApprove(address(deployment.diamond), amountToTransfer * 100);
+        (IDeal deal,) = deployment.deployDeal(
+            TestHelper.DeployDealParams({depositAmount: amountToTransfer, dealParams: dealParams})
         );
-        vm.startPrank(address(deployment.diamond));
-        deal.addComputeUnit(computeProvider, unitId, peerId);
+
+        // Match the Deal with the Offers
+        bytes32[][] memory unitIdsByOffers = TestHelper.flattenUnitIdsToOffers(registeredOffers);
+        deployment.diamondAsMarket.matchDeal(IDeal(address(deal)), registeredOffers.offerIds, unitIdsByOffers);
```
Comment (crystalbit) [inline test/deal/Rewards.t.sol:203]:
```suggestion
        deployment.diamondAsMarket.matchDeal(deal, registeredOffers.offerIds, unitIdsByOffers);
```

File: test/deal/Rewards.t.sol [hunk 16]
```diff
@@ -24,117 +24,454 @@ import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
...
+        // Add remainig workers
+
+        uint256 remainingWorkersAmount = offerParams.peerCountPerOffer - INITIAL_WORKERS; // Works for 1 CU per Peer
+        vm.startPrank(dealMaker);
+        bytes32[][] memory NewlUnitIdsToMatcher = new bytes32[][](1);
+        NewlUnitIdsToMatcher[0] = new bytes32[](remainingWorkersAmount);
+        for (uint256 i = 0; i < remainingWorkersAmount; i++) {
+            NewlUnitIdsToMatcher[0][i] = unitIdsByOffers[0][INITIAL_WORKERS + i];
+        }
+        deployment.diamondAsMarket.matchDeal(IDeal(address(deal)), registeredOffers.offerIds, NewlUnitIdsToMatcher);
```
Comment (crystalbit) [inline test/deal/Rewards.t.sol:381]:
```suggestion
        deployment.diamondAsMarket.matchDeal(deal, registeredOffers.offerIds, NewlUnitIdsToMatcher);
```

File: test/utils/TestHelper.sol [hunk 17]
```diff
@@ -18,8 +18,8 @@
  */
 pragma solidity ^0.8.19;
 
+import "forge-std/console.sol";
```
Comment (crystalbit) [inline test/utils/TestHelper.sol:21]: del pls

**Pull request merged**